File: packages/dvai-bridge-capacitor-llama/android/src/test/java/co/deepvoiceai/dvaibridge/llama/ImageDecoderTest.kt
package co.deepvoiceai.dvaibridge.llama

import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okio.Buffer
import org.junit.After
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.File
import java.nio.file.Files

@RunWith(RobolectricTestRunner::class)
class ImageDecoderTest {
    private lateinit var server: MockWebServer

    @Before
    fun setUp() {
        server = MockWebServer().apply { start(0) }
    }

    @After
    fun tearDown() {
        server.shutdown()
    }

    /** `data:image/png;base64,...` decodes to bytes starting with the PNG magic header. */
    @Test
    fun `data URL base64 decodes to bytes`() {
        val url = loadResourceText("tiny-test-base64.txt")
        val bytes = ImageDecoder.resolve(url)
        assertArrayEquals(
            byteArrayOf(0x89.toByte(), 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A),
            bytes.copyOfRange(0, 8),
        )
    }

    /** `https://...` URLs fetch the body bytes verbatim. */
    @Test
    fun `https URL fetches bytes`() {
        val payload = loadResource("tiny-test.png")
        server.enqueue(MockResponse().setBody(Buffer().apply { write(payload) }))
        val bytes = ImageDecoder.resolveWithClient(
            server.url("/img.png").toString(),
            OkHttpClient(),
        )
        assertArrayEquals(payload, bytes)
    }

    /** `file://` URLs read the file off disk. */
    @Test
    fun `file URL reads bytes`() {
        val payload = loadResource("tiny-test.png")
        val tmp = Files.createTempFile("dvai-image-", ".png").toFile().apply {
            writeBytes(payload)
            deleteOnExit()
        }
        try {
            val bytes = ImageDecoder.resolve(tmp.toURI().toString())
            assertArrayEquals(payload, bytes)
        } finally {
            tmp.delete()
        }
    }

    /** Unsupported scheme → `InvalidScheme`. */
    @Test(expected = ImageSourceError.InvalidScheme::class)
    fun `unsupported scheme throws`() {
        ImageDecoder.resolve("ftp://example.com/x.png")
    }

    /** `data:` URL with no comma → `MalformedDataURL`. */
    @Test(expected = ImageSourceError.MalformedDataURL::class)
    fun `malformed data URL throws`() {
        ImageDecoder.resolve("data:image/png;base64")
    }

    /** `file:` URL with no path → `MalformedURL` (was previously misreported as `InvalidScheme`). */
    @Test(expected = ImageSourceError.MalformedURL::class)
    fun `file URL with no path throws MalformedURL`() {
        ImageDecoder.resolve("file:")
    }

    /** HTTP non-2xx → `HttpError` carrying the status code. */
    @Test
    fun `http error throws HttpError`() {
        server.enqueue(MockResponse().setResponseCode(404))
        try {
            ImageDecoder.resolveWithClient(
                server.url("/missing.png").toString(),
                OkHttpClient(),
            )
            fail("Expected ImageSourceError.HttpError")
        } catch (e: ImageSourceError.HttpError) {
            assertEquals(404, e.status)
        }
    }

    private fun loadResource(name: String): ByteArray =
        javaClass.getResourceAsStream("/images/$name")
            ?.use { it.readBytes() }
            ?: error("missing test resource: /images/$name")

    private fun loadResourceText(name: String): String =
        javaClass.getResourceAsStream("/images/$name")
            ?.bufferedReader()
            ?.use { it.readText().trim() }
            ?: error("missing test resource: /images/$name")
}

File: packages/dvai-bridge-capacitor-llama/ios/Tests/DVAICapacitorLlamaTests/ImageDecoderTest.swift
import XCTest
@testable import DVAICapacitorLlama

final class ImageDecoderTest: XCTestCase {
    /// `data:image/png;base64,...` round-trips to bytes whose first 8 bytes
    /// are the canonical PNG magic header.
    func testDataURLBase64() async throws {
        let url = try String(contentsOf: imageFixtureURL("tiny-test-base64.txt"), encoding: .utf8)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let bytes = try await ImageDecoder.resolve(url: url)
        XCTAssertEqual(
            Array(bytes.prefix(8)),
            [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A],
            "expected PNG magic header"
        )
    }

    /// `file://` URLs return the raw bytes off disk.
    func testFileURL() async throws {
        let pngURL = imageFixtureURL("tiny-test.png")
        let result = try await ImageDecoder.resolve(url: pngURL.absoluteString)
        let raw = try Data(contentsOf: pngURL)
        XCTAssertEqual(result, raw)
    }

    /// Unsupported schemes throw `invalidScheme`.
    func testInvalidScheme() async {
        do {
            _ = try await ImageDecoder.resolve(url: "ftp://example.com/x.png")
            XCTFail("Expected throw")
        } catch ImageSourceError.invalidScheme {
            // expected
        } catch {
            XCTFail("Unexpected error type: \(error)")
        }
    }

    /// `data:` URL with no comma → `malformedDataURL`.
    func testMalformedDataURL() async {
        do {
            _ = try await ImageDecoder.resolve(url: "data:image/png;base64")
            XCTFail("Expected throw")
        } catch ImageSourceError.malformedDataURL {
            // expected
        } catch {
            XCTFail("Unexpected error type: \(error)")
        }
    }

    /// `https://` URL fetches response body bytes verbatim. Mocked at the
    /// URLSession layer via `URLProtocol.registerClass` so no real network
    /// is touched.
    func testHTTPSFetchesBytes() async throws {
        let payload = try Data(contentsOf: imageFixtureURL("tiny-test.png"))
        URLProtocol.registerClass(MockURLProtocol.self)
        defer {
            URLProtocol.unregisterClass(MockURLProtocol.self)
            MockURLProtocol.handler = nil
        }
        MockURLProtocol.handler = { request in
            let response = HTTPURLResponse(
                url: request.url!,
                statusCode: 200,
                httpVersion: "HTTP/1.1",
                headerFields: nil
            )!
            return (response, payload)
        }

        let bytes = try await ImageDecoder.resolve(url: "https://example.invalid/img.png")
        XCTAssertEqual(bytes, payload)
    }

    /// HTTP non-2xx → `ImageSourceError.httpError(status:)` carrying the code.
    func testHTTPErrorThrowsHttpError() async {
        URLProtocol.registerClass(MockURLProtocol.self)
        defer {
            URLProtocol.unregisterClass(MockURLProtocol.self)
            MockURLProtocol.handler = nil
        }
        MockURLProtocol.handler = { request in
            let response = HTTPURLResponse(
                url: request.url!,
                statusCode: 404,
                httpVersion: "HTTP/1.1",
                headerFields: nil
            )!
            return (response, Data())
        }

        do {
            _ = try await ImageDecoder.resolve(url: "https://example.invalid/missing.png")
            XCTFail("Expected throw")
        } catch ImageSourceError.httpError(let status) {
            XCTAssertEqual(status, 404)
        } catch {
            XCTFail("Unexpected error type: \(error)")
        }
    }

    /// Walks up from this test source file until it finds the repo-root
    /// `fixtures/` directory — same pattern as `AudioDecoderTest`.
    private func imageFixtureURL(_ name: String) -> URL {
        var dir = URL(fileURLWithPath: #file).deletingLastPathComponent()
        while !FileManager.default.fileExists(atPath: dir.appendingPathComponent("fixtures").path) {
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path {
                fatalError("fixtures dir not found walking up from \(#file)")
            }
            dir = parent
        }
        return dir.appendingPathComponent("fixtures").appendingPathComponent("images").appendingPathComponent(name)
    }
}

/// In-process `URLProtocol` stub that intercepts every URLSession request
/// and dispatches it to a per-test handler. Registered globally via
/// `URLProtocol.registerClass`, which `URLSession.shared` consults — so the
/// production code under test (which uses `URLSession.shared`) is exercised
/// without any actual network I/O.
private final class MockURLProtocol: URLProtocol {
    static var handler: ((URLRequest) -> (HTTPURLResponse, Data))?

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        guard let handler = MockURLProtocol.handler else {
            client?.urlProtocol(self, didFailWithError: NSError(domain: "MockURLProtocol", code: -1))
            return
        }
        let (response, data) = handler(request)
        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
        client?.urlProtocol(self, didLoad: data)
        client?.urlProtocolDidFinishLoading(self)
    }

    override func stopLoading() {}
}
